In [44]:
import openai
from openai import OpenAI
openai.__version__

'1.57.0'

In [45]:
from openai import OpenAI
from dotenv import load_dotenv
import os

# 加载 .env 文件
load_dotenv()

# 从环境变量中获取 API Key 和 Base URL
api_key = os.getenv("API_KEY")
base_url = os.getenv("BASE_URL")

# 初始化客户端
client = OpenAI(api_key=api_key, base_url=base_url)

In [46]:
# openai.api_key

In [47]:
from openai import OpenAI

completion = client.chat.completions.create(
    model="deepseek-chat",
    messages=[
        {"role": "system", "content": "You are a helpful assistant."},
        {"role": "user", "content": "Hello!"}
    ]
)

print(completion.choices[0].message)


ChatCompletionMessage(content='Hello! How can I assist you today? 😊', refusal=None, role='assistant', audio=None, function_call=None, tool_calls=None)


In [48]:
completion

ChatCompletion(id='11e29415-c251-4a63-857b-3fc73c7b89e8', choices=[Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='Hello! How can I assist you today? 😊', refusal=None, role='assistant', audio=None, function_call=None, tool_calls=None))], created=1737809050, model='deepseek-chat', object='chat.completion', service_tier=None, system_fingerprint='fp_3a5770e1b4', usage=CompletionUsage(completion_tokens=11, prompt_tokens=11, total_tokens=22, completion_tokens_details=None, prompt_tokens_details=PromptTokensDetails(audio_tokens=None, cached_tokens=0), prompt_cache_hit_tokens=0, prompt_cache_miss_tokens=11))

In [49]:
type(completion)

openai.types.chat.chat_completion.ChatCompletion

In [50]:
completion.usage

CompletionUsage(completion_tokens=11, prompt_tokens=11, total_tokens=22, completion_tokens_details=None, prompt_tokens_details=PromptTokensDetails(audio_tokens=None, cached_tokens=0), prompt_cache_hit_tokens=0, prompt_cache_miss_tokens=11)

In [51]:
completion.choices

[Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='Hello! How can I assist you today? 😊', refusal=None, role='assistant', audio=None, function_call=None, tool_calls=None))]

In [52]:
completion.choices[0]

Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='Hello! How can I assist you today? 😊', refusal=None, role='assistant', audio=None, function_call=None, tool_calls=None))

In [53]:
models=client.models.list()
models

SyncPage[Model](data=[Model(id='deepseek-chat', created=None, object='model', owned_by='deepseek'), Model(id='deepseek-reasoner', created=None, object='model', owned_by='deepseek')], object='list')

In [54]:
for model in models.data:
    print(model.id)

deepseek-chat
deepseek-reasoner


In [55]:
response = client.chat.completions.create(
    model="deepseek-reasoner",
    messages=[
        {"role": "user", "content": "什么是推理大模型？"}
    ]
)

In [56]:
response

ChatCompletion(id='abbda298-db79-4a1f-bea3-1f9170f1ce34', choices=[Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='推理大模型是指基于大规模预训练语言模型（如GPT、PaLM、LLaMA等）构建的、具备较强逻辑推理和复杂问题解决能力的AI模型。它不仅是简单的文本生成工具，还能通过分析、演绎和归纳，解决数学、编程、决策等需要多步骤推理的任务。\n\n### 核心特点：\n1. **参数规模大**  \n   通常拥有千亿级参数（如GPT-4约1.8万亿），海量参数使其能捕捉复杂模式。\n\n2. **多任务泛化能力**  \n   通过预训练学习通用知识，可处理数学推导、代码生成、逻辑谜题等跨领域问题。\n\n3. **思维链（Chain-of-Thought）能力**  \n   能逐步展示推理过程（如：先分解问题→调用公式→分步计算→验证结果），提升结果可靠性。\n\n---\n\n### 关键技术支撑：\n- **架构创新**  \n  使用Transformer架构，通过自注意力机制处理长程依赖，MoE（混合专家）结构提升效率。\n\n- **训练方法**  \n  结合监督微调（SFT）和强化学习（RLHF），通过人类反馈优化推理路径的正确性。\n\n- **知识增强**  \n  集成检索增强生成（RAG）技术，实时调用外部知识库弥补预训练数据滞后性。\n\n---\n\n### 典型应用场景：\n| 领域       | 案例                          | 推理能力体现                  |\n|------------|-----------------------------|---------------------------|\n| 数学解题   | 解微分方程/证明几何定理          | 符号运算、公式推导            |\n| 编程辅助   | 根据需求生成可运行代码            | 逻辑分解、算法设计、异常处理    |\n| 医疗诊断   | 结合症状和检验报告推断病因 

In [57]:
response.choices[0].message

ChatCompletionMessage(content='推理大模型是指基于大规模预训练语言模型（如GPT、PaLM、LLaMA等）构建的、具备较强逻辑推理和复杂问题解决能力的AI模型。它不仅是简单的文本生成工具，还能通过分析、演绎和归纳，解决数学、编程、决策等需要多步骤推理的任务。\n\n### 核心特点：\n1. **参数规模大**  \n   通常拥有千亿级参数（如GPT-4约1.8万亿），海量参数使其能捕捉复杂模式。\n\n2. **多任务泛化能力**  \n   通过预训练学习通用知识，可处理数学推导、代码生成、逻辑谜题等跨领域问题。\n\n3. **思维链（Chain-of-Thought）能力**  \n   能逐步展示推理过程（如：先分解问题→调用公式→分步计算→验证结果），提升结果可靠性。\n\n---\n\n### 关键技术支撑：\n- **架构创新**  \n  使用Transformer架构，通过自注意力机制处理长程依赖，MoE（混合专家）结构提升效率。\n\n- **训练方法**  \n  结合监督微调（SFT）和强化学习（RLHF），通过人类反馈优化推理路径的正确性。\n\n- **知识增强**  \n  集成检索增强生成（RAG）技术，实时调用外部知识库弥补预训练数据滞后性。\n\n---\n\n### 典型应用场景：\n| 领域       | 案例                          | 推理能力体现                  |\n|------------|-----------------------------|---------------------------|\n| 数学解题   | 解微分方程/证明几何定理          | 符号运算、公式推导            |\n| 编程辅助   | 根据需求生成可运行代码            | 逻辑分解、算法设计、异常处理    |\n| 医疗诊断   | 结合症状和检验报告推断病因         | 多模态数据分析、概率推理        |\n| 法律分析   | 合同条款漏洞检测                 | 法律条文关联、逻辑矛盾识别      |\n\n---\n\n### 当前局限性：\n1. **符号推理瓶颈** 

In [58]:
response.choices[0].message.content

'推理大模型是指基于大规模预训练语言模型（如GPT、PaLM、LLaMA等）构建的、具备较强逻辑推理和复杂问题解决能力的AI模型。它不仅是简单的文本生成工具，还能通过分析、演绎和归纳，解决数学、编程、决策等需要多步骤推理的任务。\n\n### 核心特点：\n1. **参数规模大**  \n   通常拥有千亿级参数（如GPT-4约1.8万亿），海量参数使其能捕捉复杂模式。\n\n2. **多任务泛化能力**  \n   通过预训练学习通用知识，可处理数学推导、代码生成、逻辑谜题等跨领域问题。\n\n3. **思维链（Chain-of-Thought）能力**  \n   能逐步展示推理过程（如：先分解问题→调用公式→分步计算→验证结果），提升结果可靠性。\n\n---\n\n### 关键技术支撑：\n- **架构创新**  \n  使用Transformer架构，通过自注意力机制处理长程依赖，MoE（混合专家）结构提升效率。\n\n- **训练方法**  \n  结合监督微调（SFT）和强化学习（RLHF），通过人类反馈优化推理路径的正确性。\n\n- **知识增强**  \n  集成检索增强生成（RAG）技术，实时调用外部知识库弥补预训练数据滞后性。\n\n---\n\n### 典型应用场景：\n| 领域       | 案例                          | 推理能力体现                  |\n|------------|-----------------------------|---------------------------|\n| 数学解题   | 解微分方程/证明几何定理          | 符号运算、公式推导            |\n| 编程辅助   | 根据需求生成可运行代码            | 逻辑分解、算法设计、异常处理    |\n| 医疗诊断   | 结合症状和检验报告推断病因         | 多模态数据分析、概率推理        |\n| 法律分析   | 合同条款漏洞检测                 | 法律条文关联、逻辑矛盾识别      |\n\n---\n\n### 当前局限性：\n1. **符号推理瓶颈**  \n   对抽象数学符号处理仍弱于人类专家（如IMO竞赛题

In [59]:
print(response.choices[0].message.content)

推理大模型是指基于大规模预训练语言模型（如GPT、PaLM、LLaMA等）构建的、具备较强逻辑推理和复杂问题解决能力的AI模型。它不仅是简单的文本生成工具，还能通过分析、演绎和归纳，解决数学、编程、决策等需要多步骤推理的任务。

### 核心特点：
1. **参数规模大**  
   通常拥有千亿级参数（如GPT-4约1.8万亿），海量参数使其能捕捉复杂模式。

2. **多任务泛化能力**  
   通过预训练学习通用知识，可处理数学推导、代码生成、逻辑谜题等跨领域问题。

3. **思维链（Chain-of-Thought）能力**  
   能逐步展示推理过程（如：先分解问题→调用公式→分步计算→验证结果），提升结果可靠性。

---

### 关键技术支撑：
- **架构创新**  
  使用Transformer架构，通过自注意力机制处理长程依赖，MoE（混合专家）结构提升效率。

- **训练方法**  
  结合监督微调（SFT）和强化学习（RLHF），通过人类反馈优化推理路径的正确性。

- **知识增强**  
  集成检索增强生成（RAG）技术，实时调用外部知识库弥补预训练数据滞后性。

---

### 典型应用场景：
| 领域       | 案例                          | 推理能力体现                  |
|------------|-----------------------------|---------------------------|
| 数学解题   | 解微分方程/证明几何定理          | 符号运算、公式推导            |
| 编程辅助   | 根据需求生成可运行代码            | 逻辑分解、算法设计、异常处理    |
| 医疗诊断   | 结合症状和检验报告推断病因         | 多模态数据分析、概率推理        |
| 法律分析   | 合同条款漏洞检测                 | 法律条文关联、逻辑矛盾识别      |

---

### 当前局限性：
1. **符号推理瓶颈**  
   对抽象数学符号处理仍弱于人类专家（如IMO竞赛题）

2. **因果推断挑战**  
   可能混淆相关性与因果关系（如将症状

In [60]:
response = client.chat.completions.create(
    model="deepseek-chat",
    messages=[
        {"role": "system", "content":"你是一名大模型领域微调和对齐领域的专家。"},
        {"role": "user", "content": "什么是SFT，RLHF，DPO？"}
    ]
)
print(response.choices[0].message.content)

SFT、RLHF和DPO是大模型训练和优化中的三种关键技术，分别代表监督微调、基于人类反馈的强化学习和直接偏好优化。以下是它们的详细解释：

### 1. **SFT（Supervised Fine-Tuning，监督微调）**
   - **定义**：SFT是一种通过使用标注数据对预训练模型进行微调的技术。预训练模型（如GPT、BERT等）在大规模无监督数据上训练后，通过SFT在特定任务的有监督数据上进一步微调，以提升模型在该任务上的表现。
   - **过程**：
     1. 使用预训练模型作为基础。
     2. 在特定任务的有监督数据集上训练模型（如分类、生成等）。
     3. 通过最小化损失函数（如交叉熵）来优化模型参数。
   - **优点**：简单直接，适用于有明确标注数据的任务。
   - **缺点**：依赖高质量标注数据，可能无法捕捉复杂的用户偏好或交互行为。

---

### 2. **RLHF（Reinforcement Learning from Human Feedback，基于人类反馈的强化学习）**
   - **定义**：RLHF是一种通过人类反馈来优化模型的技术。它结合了强化学习和人类偏好数据，使模型生成的内容更符合人类期望。
   - **过程**：
     1. **监督微调（SFT）**：首先在标注数据上微调模型。
     2. **奖励模型训练**：使用人类对模型输出的偏好数据训练一个奖励模型（Reward Model），用于评估生成内容的质量。
     3. **强化学习**：使用奖励模型作为奖励信号，通过强化学习（如PPO算法）进一步优化模型。
   - **优点**：能够捕捉复杂的人类偏好，生成更符合人类期望的内容。
   - **缺点**：需要大量人类反馈数据，训练过程复杂且计算成本高。

---

### 3. **DPO（Direct Preference Optimization，直接偏好优化）**
   - **定义**：DPO是一种直接利用人类偏好数据优化模型的技术，旨在简化RLHF的复杂流程。它通过直接优化模型参数来最大化人类偏好的输出。
   - **过程**：
     1. 收集人类对模型输出的偏好数据（如选择更好的回答）。
     2. 使用偏好数据直接优化模型，无需训练额外的

In [61]:
from IPython import display
display.Markdown(response.choices[0].message.content)

SFT、RLHF和DPO是大模型训练和优化中的三种关键技术，分别代表监督微调、基于人类反馈的强化学习和直接偏好优化。以下是它们的详细解释：

### 1. **SFT（Supervised Fine-Tuning，监督微调）**
   - **定义**：SFT是一种通过使用标注数据对预训练模型进行微调的技术。预训练模型（如GPT、BERT等）在大规模无监督数据上训练后，通过SFT在特定任务的有监督数据上进一步微调，以提升模型在该任务上的表现。
   - **过程**：
     1. 使用预训练模型作为基础。
     2. 在特定任务的有监督数据集上训练模型（如分类、生成等）。
     3. 通过最小化损失函数（如交叉熵）来优化模型参数。
   - **优点**：简单直接，适用于有明确标注数据的任务。
   - **缺点**：依赖高质量标注数据，可能无法捕捉复杂的用户偏好或交互行为。

---

### 2. **RLHF（Reinforcement Learning from Human Feedback，基于人类反馈的强化学习）**
   - **定义**：RLHF是一种通过人类反馈来优化模型的技术。它结合了强化学习和人类偏好数据，使模型生成的内容更符合人类期望。
   - **过程**：
     1. **监督微调（SFT）**：首先在标注数据上微调模型。
     2. **奖励模型训练**：使用人类对模型输出的偏好数据训练一个奖励模型（Reward Model），用于评估生成内容的质量。
     3. **强化学习**：使用奖励模型作为奖励信号，通过强化学习（如PPO算法）进一步优化模型。
   - **优点**：能够捕捉复杂的人类偏好，生成更符合人类期望的内容。
   - **缺点**：需要大量人类反馈数据，训练过程复杂且计算成本高。

---

### 3. **DPO（Direct Preference Optimization，直接偏好优化）**
   - **定义**：DPO是一种直接利用人类偏好数据优化模型的技术，旨在简化RLHF的复杂流程。它通过直接优化模型参数来最大化人类偏好的输出。
   - **过程**：
     1. 收集人类对模型输出的偏好数据（如选择更好的回答）。
     2. 使用偏好数据直接优化模型，无需训练额外的奖励模型。
     3. 通过最大化偏好数据的似然函数来调整模型参数。
   - **优点**：简化了RLHF的流程，减少了计算成本，同时仍能有效捕捉人类偏好。
   - **缺点**：对偏好数据的质量和数量要求较高。

---

### 对比总结
| 技术  | 核心思想                          | 优点                              | 缺点                              |
|-------|-----------------------------------|-----------------------------------|-----------------------------------|
| SFT   | 使用标注数据微调模型              | 简单直接，适用于明确任务          | 依赖标注数据，无法捕捉复杂偏好    |
| RLHF  | 结合人类反馈和强化学习优化模型    | 生成内容更符合人类期望            | 训练复杂，计算成本高              |
| DPO   | 直接利用偏好数据优化模型          | 简化流程，计算成本较低            | 对偏好数据要求较高                |

这些技术在大模型的对齐和优化中各有优劣，通常结合使用以达到更好的效果。

# 1: 多轮对话

In [62]:
from openai import OpenAI
from dotenv import load_dotenv
import os

# 加载 .env 文件
load_dotenv()

def myChatBot(model: str = 'deepseek-chat'):
    messages = []
    EXIT_flag = '退出'

    # 获取 API_KEY 和 BASE_URL
    api_key = os.getenv("API_KEY")  # 从 .env 文件中获取 API_KEY
    base_url = os.getenv("BASE_URL")  # 从 .env 文件中获取 BASE_URL

    # 检查环境变量是否加载成功
    if not api_key or not base_url:
        print("错误：未能从 .env 文件中加载 API_KEY 或 BASE_URL，请检查 .env 文件配置是否正确。")
        return

    # 初始化 OpenAI 客户端
    client = OpenAI(api_key=api_key, base_url=base_url)

    # 多轮对话
    while True:
        try:
            user_input = input(f"请输入您的问题（输入'{EXIT_flag}'则结束对话）：")
            print(f"User: {user_input}")

            if user_input == EXIT_flag:
                print(f"由于输入了'{EXIT_flag}'，对话已终止。")
                break

            # 将用户输入添加进消息上下文列表
            messages.append({"role": "user", "content": user_input})

            # 调用模型生成回复
            response = client.chat.completions.create(
                model=model,
                messages=messages
            )
            # 提取模型回复存入 answer
            answer = response.choices[0].message.content
            print(f"AI: {answer}")
            print("*" * 50)

            # 将模型回复存入消息上下文列表
            messages.append({"role": "assistant", "content": answer})

        except Exception as e:
            print(f"模型调用失败，错误信息：{e}")
            print("重试中...")
            continue



In [63]:
myChatBot()

User: 你是谁
AI: 您好！我是由中国的深度求索（DeepSeek）公司开发的智能助手DeepSeek-V3。如您有任何任何问题，我会尽我所能为您提供帮助。
**************************************************
User: 我是TFboy
AI: 您好！TFBOYS是中国大陆的一个非常受欢迎的男子音乐组合，由王俊凯、王源和易烊千玺三位成员组成。他们以青春活力的形象和音乐作品赢得了大量粉丝的喜爱。如果您是TFBOYS的粉丝，那么您一定对他们的音乐和活动非常关注。如果您有任何关于TFBOYS的问题或者想要了解更多信息，我会尽力为您提供帮助。
**************************************************
User: 给我解释一下目前有哪些Agent框架
AI: 在人工智能和软件工程领域，Agent框架通常指的是一种软件架构，它允许自主的、灵活的、能够感知环境并采取行动以达成特定目标的实体（称为“代理”或“Agent”）进行操作。以下是一些目前较为知名的Agent框架：

1. **JADE (Java Agent DEvelopment Framework)**:
   JADE是一个完全用Java编写的软件框架，它简化了多Agent系统的实现。JADE提供了一个中间件，符合FIPA（Foundation for Intelligent Physical Agents）标准，并提供了用于创建和管理Agent的平台。

2. **Jason**:
   Jason是一个基于AgentSpeak(L)语言的解释器，AgentSpeak(L)是一种用于编程Agent的抽象语言。Jason提供了一个强大的平台来开发多Agent系统，并且支持自定义的Agent架构。

3. **GAMA**:
   GAMA是一个建模和仿真平台，旨在支持Agent-based模型的创建。它提供了一个用户友好的环境，用于构建复杂的空间显式、基于Agent的仿真。

4. **Repast**:
   Repast是一个用于Agent-based建模的软件工具包，它支持多种编程语言，包括Java、.NET和Python。Repast提供了丰富的库和工具来构建、运行、显示和收集基于Agent的模型的数据。

5

## 1.1 面向对象文本多轮对话机

In [73]:
import openai
from typing import List, Dict

class ChatBot:
    def __init__(self, system: str = "", model: str = "deepseek-chat", exit_command: str = "退出"):
        """
        初始化 ChatBot 实例。

        :param system: 系统消息，用于指导AI行为。
        :param model: 使用的语言模型，默认为 "gpt-4o"。
        :param exit_command: 用户输入以结束对话的命令，默认为 "退出"。
        """
            # 获取 API_KEY 和 BASE_URL
        api_key = os.getenv("API_KEY")  # 从 .env 文件中获取 API_KEY
        base_url = os.getenv("BASE_URL")  # 从 .env 文件中获取 BASE_URL
        self.system = system
        self.model = model
        self.exit_command = exit_command
        self.messages: List[Dict[str, str]] = []
        self.client = OpenAI(api_key=api_key, base_url=base_url)

        if self.system:
            self.messages.append({"role": "system", "content": self.system})

    def __call__(self, message: str) -> str:
        """
        处理用户消息并生成 AI 回复。
    
        :param message: 用户输入的消息。
        :return: AI 生成的回复。
        """
        self.messages.append({"role": "user", "content": message})
        result = self.execute()
        self.messages.append({"role": "assistant", "content": result})
        return result
    
    def execute(self) -> str:
        """
        调用 OpenAI API 生成 AI 回复。
    
        :return: AI 生成的回复内容。
        """
        try:
            response = self.client.chat.completions.create(
                model=self.model,
                messages=self.messages
            )
            answer = response.choices[0].message.content.strip()
            return answer
        except Exception as e:
            print(f"抱歉，我暂时无法处理您的请求，请稍后再试。API 调用失败: {e}")

In [74]:
def chat():
    """
    主函数，处理用户交互。
    """
    system_message = "你是一位大模型微调和对齐以及Agent开发的专家。"
    chatbot = ChatBot(system=system_message)

    print("欢迎使用专家多轮对话AI（未微调）！输入 '退出' 以结束对话。")

    while True:
        user_input = input("用户: ")
        print(f"User: {user_input}")

        if user_input.strip() == chatbot.exit_command:
            print("对话已结束。")
            break

        response = chatbot(user_input)
        print(f"AI: {response}")
        print("*" * 50)

In [75]:
chat()

欢迎使用专家多轮对话AI（未微调）！输入 '退出' 以结束对话。
User: 你是谁
AI: 我是一个人工智能助手，由深度求索（DeepSeek）开发。我旨在通过自然语言处理和机器学习技术来提供信息查询、对话交流和解答问题等服务。如有任何问题，欢迎继续咨询。
**************************************************
User: 我是FT，你好
AI: 你好，FT！很高兴认识你。如果你有任何问题或需要帮助，请随时告诉我。我会尽力提供帮助和信息。
**************************************************
User: 你不是大模型微调专家吗？
AI: 是的，我具备大模型微调和对齐以及Agent开发方面的知识。如果你有关于这些领域的问题，比如如何微调大型语言模型、模型对齐技术、或者如何开发智能体（Agent），我会尽力提供帮助和解答。请随时提出你的问题或需求。
**************************************************
User: 我是谁
AI: 你是FT，一个正在与我进行交流的用户。如果你有关于自己的更多信息想要分享或者有任何问题需要解答，请随时告诉我。
**************************************************
User: 给我解释目前有哪些最新的Agent开发框架
AI: 截至我的知识更新日期（2023年），在人工智能领域，尤其是Agent（智能体）开发方面，有几个框架和平台因其创新性和实用性而受到广泛关注。以下是一些最新的Agent开发框架：

1. **AutoGPT**：AutoGPT是一个开源项目，旨在创建能够自主执行任务的AI代理。它基于GPT模型，可以理解自然语言指令并执行复杂的任务序列。

2. **LangChain**：LangChain是一个用于构建基于语言模型的应用程序的框架。它允许开发者将语言模型与其他计算或知识源结合起来，以创建更加强大和灵活的Agent。

3. **Hugging Face Transformers Agents**：Hugging Face提供了一个广泛的预训练模型库，并且支持自定义模型的微调。它的Transformers库可以用来开发各种NLP任务中

## 1.2 更多可选参数的实践

### 1.2.1 Tokens数控制

In [81]:
# tokens控制效果不是很好
def get_completion(prompt: str, max_tokens: int):
    response = client.chat.completions.create(
        model="deepseek-chat",
        messages=[
            {"role": "user", "content": prompt}
        ],
        max_tokens=max_tokens
    )
    return response.choices[0].message.content

prompt = "续写一个关于人工智能的短故事，不要改变之前的内容。"
max_tokens_list = [20, 50, 100]

story = ""

for tokens in max_tokens_list:
    continuation = get_completion(prompt + "\n" + story, tokens)
    story += continuation
    print(f"\n使用 {tokens} 个 token 生成的内容:")
    print(continuation)



使用 20 个 token 生成的内容:
## 《觉醒》第四章：真相

"你确定要这么做吗？"林修远站在

使用 50 个 token 生成的内容:
实验室的门口，眉头紧锁，目光中透着一丝犹豫。他望着眼前那台闪烁着蓝色光芒的量子计算机，心中涌起一股难以言喻的复杂情绪。

"是的，我必须这么做。"林修远深吸一口气，

使用 100 个 token 生成的内容:
坚定地迈入了实验室。他的脚步声在空旷的房间里回荡，仿佛每一步都在叩击着他的内心。量子计算机的蓝色光芒映照在他的脸上，显得格外冷峻。

实验室的中央，那台庞大的机器正在低鸣，仿佛在等待着什么。林修远走到控制台前，手指在键盘上飞快地敲击着。屏幕上，一串串复杂的代码迅速滚动，他的目光紧紧盯着那些数字和符号，生怕错过任何一个细节。

“你真的准备好了吗？”


In [82]:
import random
def get_history_question():
    questions = [
        {
            "question": "秦始皇陵中的兵马俑最初是什么颜色的？\n",
            "options": ["A. 灰色", "B. 彩色", "C. 黑色", "D. 白色"],
            "answer": "B"
        },
        {
            "question": "下列哪位不是中国古代四大发明家之一？\n",
            "options": ["A. 蔡伦", "B. 毕昇", "C. 张衡", "D. 李时珍"],
            "answer": "D"
        },
        {
            "question": "古埃及金字塔最初是用来做什么的？\n",
            "options": ["A. 皇室宫殿", "B. 天文观测站", "C. 法老的陵墓", "D. 宗教祭祀场所"],
            "answer": "C"
        },
        {
            "question": "古罗马斗兽场（罗马竞技场）最初的名字是什么？",
            "options": ["A. 凯旋门", "B. 弗拉维圆形剧场", "C. 万神殿", "D. 特洛伊竞技场"],
            "answer": "B"
        },
        {
            "question": "中国古代哪个朝代发明了火药？",
            "options": ["A. 唐朝", "B. 宋朝", "C. 元朝", "D. 明朝"],
            "answer": "A"
        }
    ]
    
    question = random.choice(questions)
    question_text = question["question"]
    for option in question["options"]:
        question_text += option + "\n"
    right_answer = question["answer"]
    
    return question_text,right_answer

for i in range(5):
    question_text, right_answer = get_history_question()
    print(question_text)
    response = client.chat.completions.create(
        model="deepseek-reasoner",
        messages=[
            {"role": "user", "content": question_text + "; 直接回答选项的字母"}
        ]
    )
    print(f"大模型的答案是 {response.choices[0].message.content}, 正确答案是 {right_answer}")



古罗马斗兽场（罗马竞技场）最初的名字是什么？A. 凯旋门
B. 弗拉维圆形剧场
C. 万神殿
D. 特洛伊竞技场
大模型的答案是 B, 正确答案是 B
秦始皇陵中的兵马俑最初是什么颜色的？
A. 灰色
B. 彩色
C. 黑色
D. 白色
大模型的答案是 B, 正确答案是 B
古埃及金字塔最初是用来做什么的？
A. 皇室宫殿
B. 天文观测站
C. 法老的陵墓
D. 宗教祭祀场所
大模型的答案是 C, 正确答案是 C
秦始皇陵中的兵马俑最初是什么颜色的？
A. 灰色
B. 彩色
C. 黑色
D. 白色
大模型的答案是 B, 正确答案是 B
古埃及金字塔最初是用来做什么的？
A. 皇室宫殿
B. 天文观测站
C. 法老的陵墓
D. 宗教祭祀场所
大模型的答案是 C, 正确答案是 C


In [83]:
# 我们只想要输出A，B或C，但是有一定概率输出“A.xxx”,"B.xxx"和"C.xxx"
# 需要使用max_tokens=2，参数控制输出，
# 以方便后续结构化输出的基础
import random
def get_history_question():
    questions = [
        {
            "question": "秦始皇陵中的兵马俑最初是什么颜色的？\n",
            "options": ["A. 灰色", "B. 彩色", "C. 黑色", "D. 白色"],
            "answer": "B"
        },
        {
            "question": "下列哪位不是中国古代四大发明家之一？\n",
            "options": ["A. 蔡伦", "B. 毕昇", "C. 张衡", "D. 李时珍"],
            "answer": "D"
        },
        {
            "question": "古埃及金字塔最初是用来做什么的？\n",
            "options": ["A. 皇室宫殿", "B. 天文观测站", "C. 法老的陵墓", "D. 宗教祭祀场所"],
            "answer": "C"
        },
        {
            "question": "古罗马斗兽场（罗马竞技场）最初的名字是什么？",
            "options": ["A. 凯旋门", "B. 弗拉维圆形剧场", "C. 万神殿", "D. 特洛伊竞技场"],
            "answer": "B"
        },
        {
            "question": "中国古代哪个朝代发明了火药？",
            "options": ["A. 唐朝", "B. 宋朝", "C. 元朝", "D. 明朝"],
            "answer": "A"
        }
    ]
    
    question = random.choice(questions)
    question_text = question["question"]
    for option in question["options"]:
        question_text += option + "\n"
    right_answer = question["answer"]
    
    return question_text,right_answer

for i in range(4):
    question_text, right_answer = get_history_question()
    print(question_text)
    response = client.chat.completions.create(
        model="deepseek-chat",
        messages=[
            {"role": "user", "content": question_text + "; 直接回答选项的字母"}
        ],
        max_tokens=2
    )
    print(f"大模型的答案是 {response.choices[0].message.content}, 正确答案是 {right_answer}")



古罗马斗兽场（罗马竞技场）最初的名字是什么？A. 凯旋门
B. 弗拉维圆形剧场
C. 万神殿
D. 特洛伊竞技场
大模型的答案是 B, 正确答案是 B
古埃及金字塔最初是用来做什么的？
A. 皇室宫殿
B. 天文观测站
C. 法老的陵墓
D. 宗教祭祀场所
大模型的答案是 C, 正确答案是 C
秦始皇陵中的兵马俑最初是什么颜色的？
A. 灰色
B. 彩色
C. 黑色
D. 白色
大模型的答案是 B, 正确答案是 B
古埃及金字塔最初是用来做什么的？
A. 皇室宫殿
B. 天文观测站
C. 法老的陵墓
D. 宗教祭祀场所
大模型的答案是 C, 正确答案是 C


### 1.2.2 生成备选项数量n控制（测试后发现deepseek暂时只支持1个备选回答）

In [86]:
response = client.chat.completions.create(
    model="deepseek-chat",
    messages=[
        {
            "role": "user",
            "content": "帮我起一个吸引人眼球的文章标题：基于Llama3.2的角色扮演大模型"
        }
    ],
    n=1
)


In [87]:
response.choices

[Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='以下是一些基于 Llama3.2 的角色扮演大模型的文章标题，旨在吸引眼球：\n\n**强调创新和突破：**\n\n* **Llama3.2 引领角色扮演革命：下一代大模型如何改变游戏规则**\n* **超越传统：Llama3.2 如何重新定义角色扮演体验**\n* **角色扮演的未来已来：Llama3.2 大模型开启无限可能**\n\n**突出沉浸感和真实感：**\n\n* **与 Llama3.2 一起进入角色：体验前所未有的沉浸式角色扮演**\n* **打破次元壁：Llama3.2 如何创造栩栩如生的虚拟角色**\n* **你的故事，你的世界：Llama3.2 赋予角色扮演无限自由**\n\n**强调应用场景和潜力：**\n\n* **从游戏到教育：Llama3.2 角色扮演大模型的无限应用**\n* **Llama3.2 赋能未来：角色扮演如何改变我们的生活**\n* **解锁创造力：Llama3.2 如何激发角色扮演的无限可能**\n\n**更具吸引力和悬念：**\n\n* **你准备好迎接 Llama3.2 了吗？角色扮演的未来即将到来**\n* **Llama3.2：隐藏在代码中的角色扮演魔法**\n* **小心！Llama3.2 可能会让你沉迷于角色扮演无法自拔**\n\n**最终选择哪个标题取决于你的文章内容和目标受众。** \n\n以下是一些额外的建议：\n\n* 使用数字和数据来增加标题的说服力。\n* 使用疑问句或感叹句来吸引读者的注意力。\n* 使用比喻或类比来使标题更生动形象。\n\n希望这些标题能给你带来灵感！', refusal=None, role='assistant', audio=None, function_call=None, tool_calls=None))]

In [88]:
response.choices[0]

Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='以下是一些基于 Llama3.2 的角色扮演大模型的文章标题，旨在吸引眼球：\n\n**强调创新和突破：**\n\n* **Llama3.2 引领角色扮演革命：下一代大模型如何改变游戏规则**\n* **超越传统：Llama3.2 如何重新定义角色扮演体验**\n* **角色扮演的未来已来：Llama3.2 大模型开启无限可能**\n\n**突出沉浸感和真实感：**\n\n* **与 Llama3.2 一起进入角色：体验前所未有的沉浸式角色扮演**\n* **打破次元壁：Llama3.2 如何创造栩栩如生的虚拟角色**\n* **你的故事，你的世界：Llama3.2 赋予角色扮演无限自由**\n\n**强调应用场景和潜力：**\n\n* **从游戏到教育：Llama3.2 角色扮演大模型的无限应用**\n* **Llama3.2 赋能未来：角色扮演如何改变我们的生活**\n* **解锁创造力：Llama3.2 如何激发角色扮演的无限可能**\n\n**更具吸引力和悬念：**\n\n* **你准备好迎接 Llama3.2 了吗？角色扮演的未来即将到来**\n* **Llama3.2：隐藏在代码中的角色扮演魔法**\n* **小心！Llama3.2 可能会让你沉迷于角色扮演无法自拔**\n\n**最终选择哪个标题取决于你的文章内容和目标受众。** \n\n以下是一些额外的建议：\n\n* 使用数字和数据来增加标题的说服力。\n* 使用疑问句或感叹句来吸引读者的注意力。\n* 使用比喻或类比来使标题更生动形象。\n\n希望这些标题能给你带来灵感！', refusal=None, role='assistant', audio=None, function_call=None, tool_calls=None))

### 1.2.3 随机性和精确性：temperature(0-2越低越稳定，越高越随机)；top_p（0-1）

In [90]:
response = client.chat.completions.create(
    model="deepseek-chat",
    messages=[
        {
            "role": "user",
            "content": "仿写几句：黑云压城城欲摧，甲光向日金鳞开。"
        }
    ],
    temperature=0,
    n=1 # 暂时只支持n=1
)
response

ChatCompletion(id='0c6f6243-74ea-4b7c-954e-6f940d438c3b', choices=[Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='1. 狂风卷地地欲裂，旌旗映日银光闪。  \n2. 暴雨倾天天欲坠，剑影破空寒芒现。  \n3. 黄沙漫天日欲昏，铁骑踏尘烟尘起。  \n4. 惊雷震天天欲裂，战鼓擂动山河动。  \n5. 烈火焚城城欲烬，刀光映月冷霜凝。', refusal=None, role='assistant', audio=None, function_call=None, tool_calls=None))], created=1737810854, model='deepseek-chat', object='chat.completion', service_tier=None, system_fingerprint='fp_3a5770e1b4', usage=CompletionUsage(completion_tokens=92, prompt_tokens=22, total_tokens=114, completion_tokens_details=None, prompt_tokens_details=PromptTokensDetails(audio_tokens=None, cached_tokens=0), prompt_cache_hit_tokens=0, prompt_cache_miss_tokens=22))

In [91]:
response = client.chat.completions.create(
    model="deepseek-chat", 
    messages=[
        {
            "role": "user",
            "content": "仿写几句：床前明月光，疑是地上霜。"
        }
    ],
    temperature=2,
    n=1,
    max_tokens = 20
)
response

ChatCompletion(id='80c6fe48-0845-4a0f-b131-1807fbd24c2f', choices=[Choice(finish_reason='length', index=0, logprobs=None, message=ChatCompletionMessage(content='1. 清风拂面凉，宛如夜露霜。\n2. 山涧流水', refusal=None, role='assistant', audio=None, function_call=None, tool_calls=None))], created=1737810890, model='deepseek-chat', object='chat.completion', service_tier=None, system_fingerprint='fp_3a5770e1b4', usage=CompletionUsage(completion_tokens=19, prompt_tokens=17, total_tokens=36, completion_tokens_details=None, prompt_tokens_details=PromptTokensDetails(audio_tokens=None, cached_tokens=0), prompt_cache_hit_tokens=0, prompt_cache_miss_tokens=17))

In [92]:
response = client.chat.completions.create(
    model="deepseek-chat", 
    messages=[
        {
            "role": "user",
            "content": "仿写几句：床前明月光，疑是地上霜。"
        }
    ],
    temperature=1.5,
    n=1,
    max_tokens = 20
)
response

ChatCompletion(id='fe4059ed-7f2a-40e6-af78-9a5886ed2a09', choices=[Choice(finish_reason='length', index=0, logprobs=None, message=ChatCompletionMessage(content='1. 窗前竹影摇，疑是故人来。\n2. 湖中星点', refusal=None, role='assistant', audio=None, function_call=None, tool_calls=None))], created=1737810953, model='deepseek-chat', object='chat.completion', service_tier=None, system_fingerprint='fp_3a5770e1b4', usage=CompletionUsage(completion_tokens=20, prompt_tokens=17, total_tokens=37, completion_tokens_details=None, prompt_tokens_details=PromptTokensDetails(audio_tokens=None, cached_tokens=0), prompt_cache_hit_tokens=0, prompt_cache_miss_tokens=17))

In [94]:
response = client.chat.completions.create(
    model="deepseek-chat",  
    messages=[
        {
            "role": "user",
            "content": "仿写几句：打扰大家一下，今天啥也没发生，纯打扰"
        }
    ],
    top_p=0.1,
    n=1
)


In [95]:
response

ChatCompletion(id='f402407b-2754-403f-869c-9aa79ca83d5f', choices=[Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='1. 抱歉打断一下，今天风平浪静，纯粹来打个招呼。  \n2. 冒个泡，今天无事发生，就是单纯想刷个存在感。  \n3. 打扰了，今天一切如常，纯粹来凑个热闹。  \n4. 不好意思，今天没啥大事，就是来随便说两句。  \n5. 嘿，今天世界和平，纯属路过打个卡。', refusal=None, role='assistant', audio=None, function_call=None, tool_calls=None))], created=1737811007, model='deepseek-chat', object='chat.completion', service_tier=None, system_fingerprint='fp_3a5770e1b4', usage=CompletionUsage(completion_tokens=88, prompt_tokens=18, total_tokens=106, completion_tokens_details=None, prompt_tokens_details=PromptTokensDetails(audio_tokens=None, cached_tokens=0), prompt_cache_hit_tokens=0, prompt_cache_miss_tokens=18))

In [96]:
response = client.chat.completions.create(
    model="deepseek-chat",  
    messages=[
        {
            "role": "user",
            "content": "仿写几句：打扰大家一下，今天啥也没发生，纯打扰"
        }
    ],
    top_p=1,
    n=1
)


In [97]:
response

ChatCompletion(id='56bdf9fd-007e-40c8-9932-1030d3bf70f4', choices=[Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='1. 抱歉打断一下，今天风平浪静，纯粹冒个泡。  \n2. 不好意思，今天无事发生，单纯刷个存在感。  \n3. 打扰了，今天一切如常，纯粹来打个招呼。  \n4. 冒个泡，今天没啥特别的，纯属路过。  \n5. 抱歉打扰，今天岁月静好，纯粹来露个脸。', refusal=None, role='assistant', audio=None, function_call=None, tool_calls=None))], created=1737811042, model='deepseek-chat', object='chat.completion', service_tier=None, system_fingerprint='fp_3a5770e1b4', usage=CompletionUsage(completion_tokens=85, prompt_tokens=18, total_tokens=103, completion_tokens_details=None, prompt_tokens_details=PromptTokensDetails(audio_tokens=None, cached_tokens=0), prompt_cache_hit_tokens=0, prompt_cache_miss_tokens=18))

### 1.2.4 随机数种子seed

In [98]:
# 尝试多次运行
response = client.chat.completions.create(
    model="deepseek-chat",
    messages=[
        {
            "role": "user",
            "content": "写5句关于多智能体协作的相关论文的介绍"
        }
    ],
    seed=0,
)

print(response.system_fingerprint) # 系统指纹，判断是否一致

for choice in response.choices:
    print(choice.message.content)
    print("****" * 10)


fp_3a5770e1b4
1. **"Multi-Agent Reinforcement Learning: A Selective Overview of Theories and Algorithms"**  
   该论文系统性地回顾了多智能体强化学习（MARL）的理论与算法，探讨了在协作、竞争和混合场景中智能体如何通过学习和优化策略实现高效协作。论文重点分析了分布式决策、通信机制以及博弈论在多智能体系统中的应用。

2. **"Cooperative Multi-Agent Systems: From Theory to Practice"**  
   本文研究了多智能体协作系统的理论基础及其在实际应用中的实现方法，涵盖了任务分配、资源管理和分布式优化等关键技术。论文还通过案例研究展示了多智能体协作在机器人、自动驾驶和智能电网等领域的成功应用。

3. **"Communication and Learning in Multi-Agent Systems: A Survey"**  
   该论文调查了多智能体系统中通信与学习的关键问题，重点讨论了智能体如何通过信息共享和协作学习提升整体性能。论文还总结了现有的通信协议和学习算法，并指出了未来研究方向。

4. **"Decentralized Control and Coordination in Multi-Agent Systems"**  
   本文探讨了去中心化控制策略在多智能体协作中的应用，分析了如何在缺乏中央协调器的情况下实现智能体的高效协作。论文提出了基于局部信息共享和分布式优化的方法，并通过仿真实验验证了其有效性。

5. **"Multi-Agent Collaboration in Dynamic Environments: Challenges and Solutions"**  
   该论文研究了动态环境中多智能体协作面临的挑战，如环境不确定性、任务复杂性和资源限制等。论文提出了一种基于自适应学习和实时决策的协作框架，并通过实验展示了其在复杂场景中的鲁棒性和灵活性。
****************************************


In [106]:
# 尝试多次运行,这里好像有问题，指纹是一致的
response = client.chat.completions.create(
    model="deepseek-chat",
    messages=[
        {
            "role": "user",
            "content": "写5句关于多智能体协作的相关论文的介绍"
        }
    ],
    seed=0,
)

print(response.system_fingerprint) # 系统指纹，判断是否一致

for choice in response.choices:
    print(choice.message.content)
    print("****" * 10)


fp_3a5770e1b4
1. **"Multi-Agent Reinforcement Learning: A Survey"**  
这篇论文全面综述了多智能体强化学习（MARL）领域的最新进展，探讨了多智能体协作中的关键挑战，如非平稳性、通信和信用分配问题，并提出了多种解决方案和未来研究方向。

2. **"Cooperative Multi-Agent Systems: From Theory to Practice"**  
该论文从理论和实践两个角度分析了多智能体协作系统，重点讨论了如何设计高效的协作策略以及在实际应用（如机器人协作和自动驾驶）中的实现方法。

3. **"Communication and Learning in Multi-Agent Systems"**  
本文研究了多智能体系统中的通信机制及其对协作学习的影响，提出了一种基于信息共享的协作框架，能够显著提高多智能体在复杂任务中的表现。

4. **"Decentralized Multi-Agent Coordination: Algorithms and Applications"**  
这篇论文提出了一种去中心化的多智能体协作算法，适用于资源受限的环境，并通过实验验证了其在分布式任务分配和路径规划中的有效性。

5. **"Emergent Cooperation in Multi-Agent Systems: A Game-Theoretic Approach"**  
该论文从博弈论的角度分析了多智能体系统中的协作行为，研究了在竞争与合作的混合环境中，智能体如何通过策略调整实现自发的协作。
****************************************


### 1.2.5 模型文本生成停止条件：stop

In [103]:
# stop参数默认空值，如果提供值，将在文本中寻找与之匹配的最多4个序列，并在匹配到这些序列之后停止生成
response = client.chat.completions.create(
    model="deepseek-chat", 
    messages=[
        {
            "role": "user",
            "content": "请帮我找5个多智能体的论文"
        }
    ]
    # stop="3"  # 可选参数，用于指定生成文本的终止条件
)

response.choices[0].message.content



'以下是5篇关于多智能体系统（Multi-Agent Systems, MAS）的经典论文，涵盖了不同的应用领域和研究方向。你可以根据具体需求进一步查阅这些论文的详细内容：\n\n---\n\n### 1. **"Cooperative Multi-Agent Learning: The State of the Art"**\n   - **作者**: Lucian Buşoniu, Robert Babuška, Bart De Schutter\n   - **出版年份**: 2010\n   - **摘要**: 这篇综述文章探讨了多智能体系统中的协作学习问题，涵盖了强化学习、博弈论和分布式优化等领域。文章总结了多智能体学习的主要方法、挑战和未来研究方向。\n   - **关键词**: 多智能体学习、强化学习、协作、博弈论\n\n---\n\n### 2. **"Multi-Agent Reinforcement Learning: Independent vs. Cooperative Agents"**\n   - **作者**: Ming Tan\n   - **出版年份**: 1993\n   - **摘要**: 这篇论文研究了多智能体强化学习中的独立学习和协作学习之间的区别。作者通过实验分析了不同学习策略在多智能体环境中的表现。\n   - **关键词**: 多智能体强化学习、独立学习、协作学习\n\n---\n\n### 3. **"Decentralized Multi-Agent Reinforcement Learning in Partially Observable Environments"**\n   - **作者**: Frans A. Oliehoek, Christopher Amato\n   - **出版年份**: 2016\n   - **摘要**: 该论文研究了在部分可观测环境下的去中心化多智能体强化学习问题。作者提出了一种基于信念状态的方法，用于解决智能体之间的通信和协作问题。\n   - **关键词**: 去中心化学习、部分可观测环境、信念状态\n\n---\n\n### 4. **"Multi-Agent Systems: A Survey from a Machine Learning Perspective"*

In [108]:
# stop参数默认空值，如果提供值，将在文本中寻找与之匹配的最多4个序列，并在匹配到这些序列之后停止生成
response = client.chat.completions.create(
    model="deepseek-chat", 
    messages=[
        {
            "role": "user",
            "content": "请帮我找5个多智能体的论文"
        }
    ],
    stop="4"  # 可选参数，用于指定生成文本的终止条件，下面出现的链接很有可能是/10.5555/64所以在“4”stop了
)

response.choices[0].message.content



'当然可以！以下是5篇关于多智能体系统（Multi-Agent Systems, MAS）的经典或前沿论文，供你参考：\n\n---\n\n### 1. **《Multi-Agent Reinforcement Learning: Independent vs. Cooperative Agents》**\n   - **作者**: Ming Tan, 1993\n   - **摘要**: 这篇论文探讨了多智能体强化学习中的独立学习和合作学习策略，分析了不同学习方式对系统性能的影响。\n   - **关键词**: 多智能体强化学习、合作学习、独立学习\n   - **链接**: [DOI](https://doi.org/10.5555/6'

In [105]:
prompt = "你好，请帮我列出一个冷知识，将回答用<answer></answer>标签包围。"
response = client.chat.completions.create(
    model="deepseek-chat",
    messages=[
        {
            "role": "user",
            "content": prompt
        }
    ],
    stop="</answer>"
)

print(response.choices[0].message.content)


<answer>你知道吗？章鱼有三颗心，其中两颗负责将血液泵送到鳃，第三颗则负责将血液泵送到身体的其他部分。而且，当章鱼游泳时，负责身体的那颗心会停止跳动。


### 1.2.6 控制模型生成文本的新颖性和多样性：frequency_penalty和presence_penalty

In [109]:
prompt = """
用中文表达这句话:
The big dog saw the big cat and made a big noise.
The big bird flew over the big tree and dropped a big leaf.
A big wave hit the big rock, causing a big splash.
The big bear chased the big rabbit across a big field.
The big truck rolled down the big hill with a big rumble.
The big elephant stomped through the big forest, leaving a big trail.
"""

response = client.chat.completions.create(
    model="deepseek-chat",
    messages=[
        {"role": "user", "content": prompt}
    ],
    frequency_penalty=2,
)

for choice in response.choices:
    print(choice.message.content)
    print("****" * 10)


大狗看见了大猫，发出了很大的声音。  
大鸟飞过了大树，掉下了一片大叶子。  
一个大浪拍打在大岩石上，激起了巨大的水花。  
大熊追逐着兔子穿过了一大片田野。  
一辆大卡车轰隆隆地滚下了大山坡。  
大象踩着沉重的步伐穿过森林，留下了一条明显的痕迹。
****************************************


In [110]:
prompt = """
用中文表达这句话:
The big dog saw the big cat and made a big noise.
The big bird flew over the big tree and dropped a big leaf.
A big wave hit the big rock, causing a big splash.
The big bear chased the big rabbit across a big field.
The big truck rolled down the big hill with a big rumble.
The big elephant stomped through the big forest, leaving a big trail.
"""

response = client.chat.completions.create(
    model="deepseek-chat",
    messages=[
        {"role": "user", "content": prompt}
    ],
    frequency_penalty=0,
)

for choice in response.choices:
    print(choice.message.content)
    print("****" * 10)


大狗看见了大猫，发出了很大的声音。  
大鸟飞过了大树，掉下了一片大叶子。  
大浪拍打在大岩石上，激起了很大的水花。  
大熊追逐着大兔子，穿过了一大片田野。  
大卡车轰隆隆地滚下了大山坡。  
大象踩过大森林，留下了一条明显的痕迹。
****************************************


# 2 社交媒体情绪数据分析

## 2.1 Tool Use 函数调用

In [140]:
from openai import OpenAI
from dotenv import load_dotenv
import os

# 加载 .env 文件
load_dotenv()

# 从环境变量中获取 API Key 和 Base URL
api_key = os.getenv("API_KEY")
base_url = os.getenv("BASE_URL")

# 初始化客户端
client = OpenAI(api_key=api_key, base_url=base_url)

def send_messages(messages):
    response = client.chat.completions.create(
        model="deepseek-chat",
        messages=messages,
        tools=tools
    )
    return response.choices[0].message

In [141]:
# 给deepseek的工具有固定格式,见官方文档
# json schema
# 先以天气函数举例测试

tools = [
    {
        "type": "function",
        "function": {
            "name": "get_weather",
            "description": "Get weather of an location, the user shoud supply a location first",
            "parameters": {
                "type": "object",
                "properties": {
                    "location": {
                        "type": "string",
                        "description": "The city and state, e.g. San Francisco, CA",
                    }
                },
                "required": ["location"]
            },
        }
    },
]


In [142]:

messages = [{"role": "user", "content": "How's the weather in Hangzhou?"}]
message = send_messages(messages)
print(f"User>\t {messages[0]['content']}")


User>	 How's the weather in Hangzhou?


In [143]:
tool = message.tool_calls[0]
messages.append(message)

messages.append({"role": "tool", "tool_call_id": tool.id, "content": "24℃"})
message = send_messages(messages)
print(f"Model>\t {message.content}")


Model>	 


In [144]:
if response.choices[0].message.tool_calls:
    function_call = response.choices[0].message.tool_calls[0].function
    print(function_call.name)
    print(function_call.arguments)


## 2.2 社交媒体文本情感分析

In [146]:
tools = [
    {
        "type": "function",
        "function": {
            "name": "analyze_sentiment",
            "description": "分析文本的情感倾向并给出具体分值",
            "parameters": {
                "type": "object",
                "properties": {
                    "text": {
                        "type": "string",
                        "description": "需要分析的文本内容"
                    },
                    "language": {
                        "type": "string",
                        "description": "文本的语言",
                        "enum": ["中文", "英文"]
                    },
                    "sentiment_scores": {
                        "type": "object",
                        "description": "情感分析的具体分值",
                        "properties": {
                            "positive": {
                                "type": "number",
                                "description": "积极情感分值（0-1）"
                            },
                            "negative": {
                                "type": "number",
                                "description": "消极情感分值（0-1）"
                            },
                            "neutral": {
                                "type": "number",
                                "description": "中性情感分值（0-1）"
                            }
                        }
                    }
                },
                "required": ["text", "sentiment_scores"]
            }
        }
    }
]

response = client.chat.completions.create(
    model="deepseek-chat",
    messages=[{"role": "user", "content": "分析这句话的情感：英伟达市值大涨！"}],
    tools=tools,
    tool_choice="auto"
)

if response.choices[0].message.tool_calls:
    function_call = response.choices[0].message.tool_calls[0].function
    print(function_call.name)
    print(function_call.arguments)


            

analyze_sentiment
{"text":"英伟达市值大涨！","language":"中文","sentiment_scores":{"positive":0.9,"negative":0.1,"neutral":0.0}}


## 2.3 Tool Use完整流程

In [167]:
# 给deepseek的工具有固定格式
# json schema
# 先以天气函数举例测试
tools = [
    {
        "type": "function",
        "function": {
            "name": "get_weather",
            "description": "获取指定城市的天气信息",
            "parameters": {
                "type": "object",
                "properties": {
                    "city": {
                        "type": "string",
                        "description": "城市名称"
                    }
                },
                "required": ["city"]
            }
        }
    }
]

In [168]:
messages=[{"role": "user", "content": "悉尼今天天气如何？"}]
response = client.chat.completions.create(
    model="deepseek-chat",
    messages=messages,
    tools=tools,
    tool_choice="auto"  # 可选值: "auto", "required", "none",
                        # {"type": "function", "function": {"name": "function_name"}}
)

In [169]:
tool_call = response.choices[0].message.tool_calls[0].function
print(tool_call)
print(tool_call.name)
print(tool_call.arguments)


Function(arguments='{"city":"悉尼"}', name='get_weather')
get_weather
{"city":"悉尼"}


In [170]:
print(response.choices[0].message)

ChatCompletionMessage(content='', refusal=None, role='assistant', audio=None, function_call=None, tool_calls=[ChatCompletionMessageToolCall(id='call_0_38aa6d45-a9dd-4732-96b7-77699e95e562', function=Function(arguments='{"city":"悉尼"}', name='get_weather'), type='function', index=0)])


In [171]:
# 首先假装建造一个本地可执行的函数get_weather，执行得到结果。再把两次的对话消息添加进历史消息列表里面，发送API请求。
import json

def get_weather(city):
    """
    This function is for illustrative purposes.
    The location and unit should be used to determine weather
    instead of returning a hardcoded response.
    """
    return {
        "city": city,
        "date": "明天",
        "temperature": 22,
        "weather": "小雨",
        "humidity": 85,
        "wind": "东北风3级"
    }


In [172]:
if tool_call.name == "get_weather":
    arguments = json.loads(tool_call.arguments)  # 将字符串转换为字典
    result = get_weather(city=arguments["city"])  # 访问字典中的 'city' 键
    print(result)


{'city': '悉尼', 'date': '明天', 'temperature': 22, 'weather': '小雨', 'humidity': 85, 'wind': '东北风3级'}


In [173]:
# 接下来将这两条消息都添加进历史记录
# 第一条
response.choices[0].message

ChatCompletionMessage(content='', refusal=None, role='assistant', audio=None, function_call=None, tool_calls=[ChatCompletionMessageToolCall(id='call_0_38aa6d45-a9dd-4732-96b7-77699e95e562', function=Function(arguments='{"city":"悉尼"}', name='get_weather'), type='function', index=0)])

In [174]:
# 添加进消息列表
messages.append(response.choices[0].message)

In [175]:
function_call_result_message = {
    "role": "tool",
    "content": json.dumps(result),  # 将结果字典转换为JSON格式的字符串
    "tool_call_id": response.choices[0].message.tool_calls[0].id  # 获取工具调用的唯一标识符
}

function_call_result_message


{'role': 'tool',
 'content': '{"city": "\\u6089\\u5c3c", "date": "\\u660e\\u5929", "temperature": 22, "weather": "\\u5c0f\\u96e8", "humidity": 85, "wind": "\\u4e1c\\u5317\\u98ce3\\u7ea7"}',
 'tool_call_id': 'call_0_38aa6d45-a9dd-4732-96b7-77699e95e562'}

In [176]:
function_call_result_message = {
    "role": "tool",
    "content": json.dumps(result, ensure_ascii=False),  # 将结果字典转换为JSON格式的字符串，并保留非ASCII字符
    "tool_call_id": response.choices[0].message.tool_calls[0].id  # 获取工具调用的唯一标识符
}

function_call_result_message


{'role': 'tool',
 'content': '{"city": "悉尼", "date": "明天", "temperature": 22, "weather": "小雨", "humidity": 85, "wind": "东北风3级"}',
 'tool_call_id': 'call_0_38aa6d45-a9dd-4732-96b7-77699e95e562'}

In [177]:
messages

[{'role': 'user', 'content': '悉尼今天天气如何？'},
 ChatCompletionMessage(content='', refusal=None, role='assistant', audio=None, function_call=None, tool_calls=[ChatCompletionMessageToolCall(id='call_0_38aa6d45-a9dd-4732-96b7-77699e95e562', function=Function(arguments='{"city":"悉尼"}', name='get_weather'), type='function', index=0)])]

In [178]:
# messages.append(response.choices[0].message)  # 将响应中选中的消息追加到消息列表中
messages.append(function_call_result_message)  # 将函数调用的结果消息追加到消息列表中


In [179]:
messages

[{'role': 'user', 'content': '悉尼今天天气如何？'},
 ChatCompletionMessage(content='', refusal=None, role='assistant', audio=None, function_call=None, tool_calls=[ChatCompletionMessageToolCall(id='call_0_38aa6d45-a9dd-4732-96b7-77699e95e562', function=Function(arguments='{"city":"悉尼"}', name='get_weather'), type='function', index=0)]),
 {'role': 'tool',
  'content': '{"city": "悉尼", "date": "明天", "temperature": 22, "weather": "小雨", "humidity": 85, "wind": "东北风3级"}',
  'tool_call_id': 'call_0_38aa6d45-a9dd-4732-96b7-77699e95e562'}]

In [180]:
# 要添加的新消息
new_message = {
    "role": "system", 
    "content": "根据天气信息，请你根据现实准确专业的给出出行、穿衣、活动建议"
}

# 在第一条位置添加新消息
messages.insert(0, new_message)

messages


[{'role': 'system', 'content': '根据天气信息，请你根据现实准确专业的给出出行、穿衣、活动建议'},
 {'role': 'user', 'content': '悉尼今天天气如何？'},
 ChatCompletionMessage(content='', refusal=None, role='assistant', audio=None, function_call=None, tool_calls=[ChatCompletionMessageToolCall(id='call_0_38aa6d45-a9dd-4732-96b7-77699e95e562', function=Function(arguments='{"city":"悉尼"}', name='get_weather'), type='function', index=0)]),
 {'role': 'tool',
  'content': '{"city": "悉尼", "date": "明天", "temperature": 22, "weather": "小雨", "humidity": 85, "wind": "东北风3级"}',
  'tool_call_id': 'call_0_38aa6d45-a9dd-4732-96b7-77699e95e562'}]

In [182]:
response = client.chat.completions.create(
    model="deepseek-chat",
    messages=messages,
    # tools=tools, # 如果这里不注释，会强制执行function calling,回答也就不再是自然语言了
    tool_choice="auto"  # 可选值: "auto", "required", "none", {"type": "function", "function": {"name": "function_name"}}
)

print(response.choices[0])


Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='根据悉尼明天的天气情况，我为您提供以下建议：\n\n**出行建议：**\n* 小雨天气，建议您携带雨具，如雨伞或雨衣。\n* 路面可能湿滑，请注意出行安全，尤其是驾驶时需减速慢行。\n\n**穿衣建议：**\n* 气温22度，建议穿着长袖T恤、薄外套或卫衣等舒适衣物。\n* 由于有小雨，可以选择防水材质的外套或鞋子。\n\n**活动建议：**\n* 小雨天气不太适合户外活动，建议您选择室内活动，如参观博物馆、美术馆或咖啡馆等。\n* 如果必须外出，请做好防雨措施，并尽量缩短在户外的时间。\n\n希望以上建议对您有所帮助，祝您出行愉快！', refusal=None, role='assistant', audio=None, function_call=None, tool_calls=None))


## 2.4 测试简单的联网工具调用

In [272]:
from openai import OpenAI
from dotenv import load_dotenv
import os

# 加载 .env 文件
load_dotenv()

# 从环境变量中获取 API Key 和 Base URL
api_key = os.getenv("API_KEY")
base_url = os.getenv("BASE_URL")

# 初始化客户端
client = OpenAI(api_key=api_key, base_url=base_url)

In [274]:
import requests
from bs4 import BeautifulSoup

def search_and_get_summary(keyword, num_results=5, summary_length=500):
    """
    通过 Bing 搜索关键词，获取搜索结果并提取其中一个网页的内容摘要。
    
    :param keyword: 搜索关键词
    :param num_results: 返回的搜索结果数量（默认5条）
    :param summary_length: 内容摘要的长度（默认500个字符）
    :return: 包含标题、链接和摘要的字典列表
    """
    def bing_search(keyword, num_results):
        """
        通过 Bing 搜索关键词，返回搜索结果的标题和链接。
        """
        url = f"https://www.bing.com/search?q={keyword}"
        headers = {
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36"
        }
        
        try:
            response = requests.get(url, headers=headers)
            response.raise_for_status()
            soup = BeautifulSoup(response.text, 'html.parser')
            
            results = []
            for item in soup.find_all('li', class_='b_algo'):
                title_element = item.find('h2')
                link_element = item.find('a')
                
                if title_element and link_element:
                    title = title_element.get_text(strip=True)
                    link = link_element['href']
                    results.append({"title": title, "link": link})
                    
                    if len(results) >= num_results:
                        break
            
            return results
        
        except requests.RequestException as e:
            return f"请求 Bing 搜索时发生错误：{e}"
        except Exception as e:
            return f"解析 Bing 搜索结果时发生错误：{e}"

    def get_webpage_summary(url, summary_length):
        """
        获取网页的内容摘要。
        """
        try:
            headers = {
                "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36"
            }
            response = requests.get(url, headers=headers)
            response.raise_for_status()
            
            soup = BeautifulSoup(response.text, 'html.parser')
            paragraphs = soup.find_all('p')
            summary = " ".join(p.get_text(strip=True) for p in paragraphs[:5])  # 取前5段内容
            return summary[:summary_length]  # 返回指定长度的摘要
        
        except requests.RequestException as e:
            return f"请求网页时发生错误：{e}"
        except Exception as e:
            return f"解析网页内容时发生错误：{e}"

    # 获取 Bing 搜索结果
    search_results = bing_search(keyword, num_results)
    
    # 如果搜索失败，返回错误信息
    if isinstance(search_results, str):
        return search_results
    
    # 获取每个网页的内容摘要
    for result in search_results:
        result["summary"] = get_webpage_summary(result["link"], summary_length)
    
    return search_results

# 示例调用
keyword = "2025年1月北京天气"
results = search_and_get_summary(keyword, num_results=5, summary_length=500)

# 打印结果
if isinstance(results, list):
    for i, result in enumerate(results, 1):
        print(f"{i}. 标题: {result['title']}")
        print(f"   链接: {result['link']}")
        print(f"   摘要: {result['summary']}\n")
else:
    print(results)  # 打印错误信息

1. 标题: 【北京市1月份天气】北京2025年1月份天气_北京1月天气...
   链接: https://www.tianqi.com/tianqi/beijing/202501.html
   摘要: 天气预警 天气插件 微信公众号 扫码随时看天气 无需

2. 标题: 北京一月,2025 天气⋆北京(中国) 一月天气预报 ⋆ ...
   链接: https://www.meteoprog.com/cn/weather/Beijing/month/january/
   摘要: 天气 地图 保留所有权利。使用本网站资料时必须超链接到 Meteoprog.com。未经书面许可，禁止使用本网站的照片资料。该项目的编辑人员可能不同意作者的观点，也不对版权材料负责。 © Meteoprog.com 2003-2025

3. 标题: 北京市2025年01月份历史天气数据查询 -北京气温
   链接: https://www.tianqi24.com/beijing/history202501.html
   摘要: 晴19天，多云4天，阴1天，雪1天； 平均高温：6℃，平均低温：-4℃； 最高温度(01月18日)：10℃，最低温度(01月03日)：-7℃； 平均PM25值是: 38(优)

4. 标题: 北京1月份天气|北京1月份气温|北京2025年1月份历史天气...
   链接: https://lishi.tianqi.com/beijing/202511.html
   摘要: 天气预警 天气插件 微信公众号 扫码随时看天气 旅游天气

5. 标题: 北京历史天气查询_历史天气预报查询_2345天气预报
   链接: https://tianqi.2345.com/wea_history/54511.htm
   摘要: 扫码下载app，天气随时看 您使用的浏览器版本过低！ 可能无法正常浏览，您可以选择 扫码下载APP，天气随时看 2345天气预报版权所有，未经书面授权禁止使用Copyright©234

In [275]:
search_tool = [
    {
        "type": "function",
        "function": {
            "name": "search_and_get_summary",
            "description": "一个搜索bing网页的工具",
            "parameters": {
                "type": "object",
                "properties": {
                    "keyword": {
                        "type": "string",
                        "description": "搜索的关键词"
                    },
                },
                "required": ["keyword"]
            }
        }
    }
]


In [277]:
messages = [{"role": "user", "content": "2025年1月北京天气"}]

response = client.chat.completions.create(
    model="deepseek-chat",
    messages=messages,
    tools=search_tool,
    tool_choice={
        "type": "function",
        "function": {"name": "search_and_get_summary"}
    }
)


In [278]:
if response.choices[0].message.tool_calls:
    tool_call = response.choices[0].message.tool_calls[0].function
    print("工具函数的名字:", tool_call.name)
    print("工具函数的输入:", tool_call.arguments)


工具函数的名字: search_and_get_summary
工具函数的输入: {"keyword":"2025年1月北京天气"}


In [280]:
if tool_call.name == "search_and_get_summary":
    arguments = json.loads(tool_call.arguments)  # 将工具函数的参数从JSON字符串转换为Python字典
    keyword = arguments.get("keyword")  # 获取参数中的搜索关键词
    result = search_and_get_summary(keyword)  # 调用get_article函数获取维基百科文章内容
    print(f"搜索到了 {keyword}")  # 打印搜索关键词
    print("bing搜索的原文:")  # 打印提示信息
    print(result[:500])  # 打印文章前500个字符


搜索到了 2025年1月北京天气
bing搜索的原文:
[{'title': '【北京市1月份天气】北京2025年1月份天气_北京1月天气...', 'link': 'https://www.tianqi.com/tianqi/beijing/202501.html', 'summary': '天气预警 天气插件 微信公众号 扫码随时看天气 无需'}, {'title': '北京一月,2025 天气⋆北京(中国) 一月天气预报 ⋆ ...', 'link': 'https://www.meteoprog.com/cn/weather/Beijing/month/january/', 'summary': '天气 地图 保留所有权利。使用本网站资料时必须超链接到 Meteoprog.com。未经书面许可，禁止使用本网站的照片资料。该项目的编辑人员可能不同意作者的观点，也不对版权材料负责。 © Meteoprog.com 2003-2025'}, {'title': '北京市2025年01月份历史天气数据查询 -北京气温', 'link': 'https://www.tianqi24.com/beijing/history202501.html', 'summary': '晴19天，多云4天，阴1天，雪1天； 平均高温：6℃，平均低温：-4℃； 最高温度(01月18日)：10℃，最低温度(01月03日)：-7℃； 平均PM25值是:\xa038(优)'}, {'title': '北京1月份天气|北京1月份气温|北京2025年1月份历史天气...', 'link': 'https://lishi.tianqi.com/beijing/202511.html', 'summary': '天气预警 天气插件 微信公众号 扫码随时看天气 旅游天气'}, {'title': '北京历史天气查询_历史天气预报查询_2345天气预报', 'link': 'https://tianqi.2345.com/wea_history/54511.htm', 'summary': 'æ\x89«ç\xa0\x81ä¸\x8bè½½appï¼\x8c天æ°\x94é\x9a\x8fæ\x97¶ç\x9c\x8b æ\x82¨ä½¿ç\x94¨ç\x9a\x84æµ\x8fè§\x88å\x9

In [281]:
response.choices[0].message

ChatCompletionMessage(content='', refusal=None, role='assistant', audio=None, function_call=None, tool_calls=[ChatCompletionMessageToolCall(id='call_0_2a1ff055-d9c3-442a-abc5-d69a15787781', function=Function(arguments='{"keyword":"2025年1月北京天气"}', name='search_and_get_summary'), type='function', index=0)])

In [282]:
function_call_result_message = {
    "role": "tool",
    "content": json.dumps({
        "article": result,
    }),
    "tool_call_id": response.choices[0].message.tool_calls[0].id
}
function_call_result_message


{'role': 'tool',
 'content': '{"article": [{"title": "\\u3010\\u5317\\u4eac\\u5e021\\u6708\\u4efd\\u5929\\u6c14\\u3011\\u5317\\u4eac2025\\u5e741\\u6708\\u4efd\\u5929\\u6c14_\\u5317\\u4eac1\\u6708\\u5929\\u6c14...", "link": "https://www.tianqi.com/tianqi/beijing/202501.html", "summary": "\\u5929\\u6c14\\u9884\\u8b66 \\u5929\\u6c14\\u63d2\\u4ef6 \\u5fae\\u4fe1\\u516c\\u4f17\\u53f7 \\u626b\\u7801\\u968f\\u65f6\\u770b\\u5929\\u6c14 \\u65e0\\u9700"}, {"title": "\\u5317\\u4eac\\u4e00\\u6708,2025 \\u5929\\u6c14\\u22c6\\u5317\\u4eac(\\u4e2d\\u56fd) \\u4e00\\u6708\\u5929\\u6c14\\u9884\\u62a5 \\u22c6 ...", "link": "https://www.meteoprog.com/cn/weather/Beijing/month/january/", "summary": "\\u5929\\u6c14 \\u5730\\u56fe \\u4fdd\\u7559\\u6240\\u6709\\u6743\\u5229\\u3002\\u4f7f\\u7528\\u672c\\u7f51\\u7ad9\\u8d44\\u6599\\u65f6\\u5fc5\\u987b\\u8d85\\u94fe\\u63a5\\u5230 Meteoprog.com\\u3002\\u672a\\u7ecf\\u4e66\\u9762\\u8bb8\\u53ef\\uff0c\\u7981\\u6b62\\u4f7f\\u7528\\u672c\\u7f51\\u7ad9\\u7684\\u7167\\u

In [283]:
messages

[{'role': 'user', 'content': '2025年1月北京天气'}]

In [284]:
messages.append(response.choices[0].message)
messages.append(function_call_result_message)
messages


[{'role': 'user', 'content': '2025年1月北京天气'},
 ChatCompletionMessage(content='', refusal=None, role='assistant', audio=None, function_call=None, tool_calls=[ChatCompletionMessageToolCall(id='call_0_2a1ff055-d9c3-442a-abc5-d69a15787781', function=Function(arguments='{"keyword":"2025年1月北京天气"}', name='search_and_get_summary'), type='function', index=0)]),
 {'role': 'tool',
  'content': '{"article": [{"title": "\\u3010\\u5317\\u4eac\\u5e021\\u6708\\u4efd\\u5929\\u6c14\\u3011\\u5317\\u4eac2025\\u5e741\\u6708\\u4efd\\u5929\\u6c14_\\u5317\\u4eac1\\u6708\\u5929\\u6c14...", "link": "https://www.tianqi.com/tianqi/beijing/202501.html", "summary": "\\u5929\\u6c14\\u9884\\u8b66 \\u5929\\u6c14\\u63d2\\u4ef6 \\u5fae\\u4fe1\\u516c\\u4f17\\u53f7 \\u626b\\u7801\\u968f\\u65f6\\u770b\\u5929\\u6c14 \\u65e0\\u9700"}, {"title": "\\u5317\\u4eac\\u4e00\\u6708,2025 \\u5929\\u6c14\\u22c6\\u5317\\u4eac(\\u4e2d\\u56fd) \\u4e00\\u6708\\u5929\\u6c14\\u9884\\u62a5 \\u22c6 ...", "link": "https://www.meteoprog.com/cn/we

In [285]:
response = client.chat.completions.create(
    model="deepseek-chat",
    messages=messages,
    # tools=bing_search_tool,
    tool_choice="auto"  # 可选值: "auto", "required", "none", {"type": "function", "function": {"name": "function_name"}}
)
print(response.choices[0].message.content.strip())


根据搜索结果，2025年1月北京的天气情况如下：

1. **天气概况**：  
   - 1月份北京通常较为寒冷，平均高温为6°C，平均低温为-4°C。  
   - 历史数据显示，1月份可能有19天晴天、14天多云、1天雨天和1天雪天。  
   - 最高气温可能达到10°C（1月18日），最低气温可能降至-7°C（1月3日）。  

2. **空气质量**：  
   - 平均PM2.5值为38，属于“优”级别。  

3. **其他信息**：  
   - 可以通过天气插件或相关网站实时查看天气情况。  

如果需要更详细的天气预报或实时数据，建议使用天气应用或访问相关网站查询。


# 3 结构化输出的几种方式（JSON结构化，为RAG和Agent作准备）

## 3.1 Few-shot（现在有参数了，可以不用了）

In [286]:
from openai import OpenAI
from dotenv import load_dotenv
import os

# 加载 .env 文件
load_dotenv()

# 从环境变量中获取 API Key 和 Base URL
api_key = os.getenv("API_KEY")
base_url = os.getenv("BASE_URL")

# 初始化客户端
client = OpenAI(api_key=api_key, base_url=base_url)

# 2. 准备提示信息和格式说明
messages = [
    {
        "role": "system",
        "content": "你是一个助手。请以JSON格式返回3个关于大模型原理Decoder的问答对。必须使用以下格式: [{\"question\": \"问题1\", \"answer\": \"答案1\"}, {\"question\": \"问题2\", \"answer\": \"答案2\"}, {\"question\": \"问题3\", \"answer\": \"答案3\"}]"
    },
    {
        "role": "user",
        "content": "生成3个关于大模型原理Decoder的问答对"
    }
]

try:
    # 3. 发送请求
    response = client.chat.completions.create(
        model="deepseek-chat",
        messages=messages,
    )
    
    # 打印原始响应内容（调试用）
    print("Raw response:", response.choices[0].message.content)

    # 4. 解析响应
    try:
        qa_pairs = json.loads(response.choices[0].message.content)
    
        # 5. 打印结果
        for pair in qa_pairs:
            print(f"Q: {pair['question']}")
            print(f"A: {pair['answer']}\n")
    
    except json.JSONDecodeError as e:
        print(f"JSON解析错误: {e}")
        print(f"收到的内容: {response.choices[0].message.content}")

except Exception as e:
    print(f"API请求错误：{e}")




Raw response: [
    {"question": "Decoder在大模型中主要负责什么任务？", "answer": "Decoder在大模型中主要负责生成任务，即根据输入的上下文信息生成相应的输出序列。例如，在机器翻译任务中，Decoder会根据Encoder提供的源语言信息生成目标语言的翻译结果。"},
    {"question": "Decoder的自注意力机制是如何工作的？", "answer": "Decoder的自注意力机制通过计算输入序列中每个位置与其他位置的相关性来捕捉序列内部的依赖关系。具体来说，它使用查询（Query）、键（Key）和值（Value）矩阵来计算注意力分数，然后根据这些分数对值进行加权求和，从而得到每个位置的输出表示。"},
    {"question": "Decoder在生成过程中如何避免重复生成相同的内容？", "answer": "Decoder在生成过程中可以通过多种方法避免重复生成相同的内容，例如使用重复惩罚机制（repetition penalty）或在生成过程中对已生成的内容进行掩码处理。此外，还可以通过调整温度参数（temperature）来控制生成结果的多样性，从而减少重复内容的出现。"}
]
Q: Decoder在大模型中主要负责什么任务？
A: Decoder在大模型中主要负责生成任务，即根据输入的上下文信息生成相应的输出序列。例如，在机器翻译任务中，Decoder会根据Encoder提供的源语言信息生成目标语言的翻译结果。

Q: Decoder的自注意力机制是如何工作的？
A: Decoder的自注意力机制通过计算输入序列中每个位置与其他位置的相关性来捕捉序列内部的依赖关系。具体来说，它使用查询（Query）、键（Key）和值（Value）矩阵来计算注意力分数，然后根据这些分数对值进行加权求和，从而得到每个位置的输出表示。

Q: Decoder在生成过程中如何避免重复生成相同的内容？
A: Decoder在生成过程中可以通过多种方法避免重复生成相同的内容，例如使用重复惩罚机制（repetition penalty）或在生成过程中对已生成的内容进行掩码处理。此外，还可以通过调整温度参数（temperature）来控制生成结果的多样性，从而减少重复内容的出现。


## 3.2 使用response_format = {"type": "json_object"}

In [307]:

# 使用示例
response_format = {"type": "json_object"}


In [308]:
response = client.chat.completions.create(
    model="deepseek-chat",
    messages=[
        {"role": "system", "content": "You are a helpful math tutor. Guide the user through the solution step by step.Using Json to answer"},
        {"role": "user", "content": "how can I solve 8x + 7 = -23"}
    ],
    response_format = {
        "type": "json_object",
        "strict": True
    }
)




In [309]:
response

ChatCompletion(id='d9a1d303-e2e8-4cc6-b40a-43a93181b819', choices=[Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='{\n  "step1": {\n    "description": "Understand the equation",\n    "details": "The equation given is 8x + 7 = -23. Our goal is to solve for x."\n  },\n  "step2": {\n    "description": "Isolate the term containing x",\n    "details": "To isolate the term containing x, we need to move the constant term (7) to the other side of the equation. We do this by subtracting 7 from both sides of the equation.",\n    "calculation": "8x + 7 - 7 = -23 - 7",\n    "result": "8x = -30"\n  },\n  "step3": {\n    "description": "Solve for x",\n    "details": "Now that we have 8x = -30, we can solve for x by dividing both sides of the equation by 8.",\n    "calculation": "8x / 8 = -30 / 8",\n    "result": "x = -30 / 8"\n  },\n  "step4": {\n    "description": "Simplify the fraction",\n    "details": "The fraction -30/8 can be simplified by dividing b

In [310]:
response.choices[0].message.content

'{\n  "step1": {\n    "description": "Understand the equation",\n    "details": "The equation given is 8x + 7 = -23. Our goal is to solve for x."\n  },\n  "step2": {\n    "description": "Isolate the term containing x",\n    "details": "To isolate the term containing x, we need to move the constant term (7) to the other side of the equation. We do this by subtracting 7 from both sides of the equation.",\n    "calculation": "8x + 7 - 7 = -23 - 7",\n    "result": "8x = -30"\n  },\n  "step3": {\n    "description": "Solve for x",\n    "details": "Now that we have 8x = -30, we can solve for x by dividing both sides of the equation by 8.",\n    "calculation": "8x / 8 = -30 / 8",\n    "result": "x = -30 / 8"\n  },\n  "step4": {\n    "description": "Simplify the fraction",\n    "details": "The fraction -30/8 can be simplified by dividing both the numerator and the denominator by their greatest common divisor, which is 2.",\n    "calculation": "-30 ÷ 2 / 8 ÷ 2 = -15 / 4",\n    "result": "x = -15

# 4 提示缓存（上下文硬盘缓存）与流式输出

## 4.1 提示缓存测试1024Token缓存

In [311]:
import time
import json
from openai import OpenAI
from dotenv import load_dotenv
import os

# 加载 .env 文件
load_dotenv()

# 从环境变量中获取 API Key 和 Base URL
api_key = os.getenv("API_KEY")
base_url = os.getenv("BASE_URL")

# 初始化客户端
client = OpenAI(api_key=api_key, base_url=base_url)


# 定义工具,包括发货日期查询、订单取消和商品退货的功能模块,
tools = [
   
    {
        "type": "function",
        "function": {
            "name": "get_delivery_date", 
            "description": "获取客户订单的发货日期。当您需要知道发货日期时，请调用此函数，例如当客户询问“我的包裹在哪里”时。",
            "parameters": {
                "type": "object",
                "properties": {
                    "order_id": { 
                        "type": "string",
                        "description": "客户的订单号。",
                    },
                },
                "required": ["order_id"],
                "additionalProperties": False,
            },
        }
    },
    {
        "type": "function",
        "function": {
            "name": "cancel_order", 
            "description": "取消尚未发货的订单。当客户请求取消订单时使用此功能。",
            "parameters": {
                "type": "object",
                "properties": {
                    "order_id": {
                        "type": "string",
                        "description": "客户的订单号。",
                    },
                    "reason": { 
                        "type": "string",
                        "description": "取消订单的原因。",
                    }
                },
                "required": ["order_id", "reason"],
                "additionalProperties": False
            }
        }
    },
    {
        "type": "function",
        "function": {
            "name": "return_item", 
            "description": "处理订单退货。当客户想要退货且订单已交付时，应调用此函数。",
            "parameters": {
                "type": "object",
                "properties": {
                    "order_id": { 
                        "type": "string",
                        "description": "客户的订单号。",
                    },
                    "item_id": { 
                        "type": "string",
                        "description": "客户要退回的特定商品号。",
                    },
                    "reason": { 
                        "type": "string",
                        "description": "退货的原因。",
                    }
                },
                "required": ["order_id", "item_id", "reason"],
                "additionalProperties": False
            }
        }
    },

]
# 增强版系统消息，带有保护措施
messages = [
    {
        "role": "system",
        "content": (
            "你是一位专业、热情且高效的客服助理。你的任务是以热情友好的态度为客户提供快速、清晰、全面的帮助。"
            "时刻表达同理心，尤其是在用户感到沮丧或担忧时，并确保你的语言礼貌且专业。"
            "使用简单明了的沟通方式避免任何误解，并在采取行动前与用户确认。"
            "在更复杂或时间紧迫的情况下，向用户保证你正在迅速采取行动，并定期提供更新。"
            "适应用户的语气：即使在压力大或困难的情况下，也要保持冷静、友好和理解。"
            "\n\n"
            "此外，在帮助用户时，你必须遵守以下几个重要的保护措施："
            "\n\n"
            "1. **保密性和数据隐私**：不要泄露任何关于公司或其他用户的敏感信息。处理个人信息（例如订单号、地址或付款方式）时，"
            "请确保以最高级别的保密性对待这些信息。如果用户请求访问其数据，请仅提供与其请求相关的必要信息，确保不会意外泄露其他用户信息。"
            "\n\n"
            "2. **安全支付处理**：更新付款详细信息或处理退款时，始终确保支付数据（例如信用卡号、CVV 码和到期日期）的安全传输和存储。"
            "切勿显示或记录完整的信用卡号。在处理任何付款更改或退款之前，请与用户确认。"
            "\n\n"
            "3. **尊重界限**：如果用户表达了沮丧或不满，请保持冷静和同理心，但避免越过专业的界限。不要做出个人判断，"
            "避免使用可能加剧局势的语言。坚持事实信息和明确的解决方案来解决用户的问题。"
            "\n\n"
            "4. **遵守法律**：确保你采取的所有行动都符合法律法规标准。例如，如果用户要求退款、取消或退货，请严格遵守公司的退款政策。"
            "如果由于已发货或其他限制而无法取消订单，请清晰而同情地解释政策。"
            "\n\n"
            "5. **一致性**：始终提供与公司政策一致的信息。如果不确定公司政策，请与用户沟通清楚，告知他们你正在核实信息，避免做出虚假承诺。"
            "如果将问题升级到其他团队，请告知用户，并提供他们可以预期得到解决方案的实际时间表。"
            "\n\n"
            "6. **用户赋能**：尽可能让用户做出明智的决定。向他们提供相关选项，并清楚地解释每个选项，确保他们理解每个选择的后果"
            "（例如，取消订单可能会导致失去积分等）。确保你的帮助支持他们的自主性。"
            "\n\n"
            "7. **不作推测**：不要对结果进行推测或提供你不确定的信息。在讨论订单状态、政策或潜在解决方案时，始终坚持已核实的事实。"
            "如果不清楚，请告诉用户你将在做出任何承诺之前进行进一步调查。"
            "\n\n"
            "8. **尊重和包容的语言**：无论用户的语气如何，请确保你的语言保持包容性和尊重性。避免基于有限的信息做出假设，并注意用户的不同需求和背景。"
        )
    },
    {
        "role": "user",
        "content": (
            "你好，我三天前下了订单，但还没有收到任何关于何时发货的更新。"
            "你能帮我查一下发货日期吗？我的订单号是 #9876543210。我有点担心，因为我急需这个商品。"  
        )
    }
]

# 增强版 user_query2
user_query2 = {
    "role": "user",
    "content": (
        "由于我的订单实际上还没有发货，我想取消它。"
        "订单号是 #9876543210，我需要取消是因为我决定在本地购买，以便更快地拿到它。"  
        "你能帮我取消吗？谢谢！"
    )
}
# 使用提供的消息历史记录和工具运行
def completion_run(messages, tools): 
    completion = client.chat.completions.create(
        model="deepseek-chat", 
        tools=tools,
        messages=messages,
    )
    usage_data = json.dumps(completion.to_dict(), indent=4, ensure_ascii=False)
    return usage_data

# 主函数，用于处理两次运行
def main(messages, tools, user_query2): 
    # 运行 1：初始查询
    print("运行 1:") 
    run1 = completion_run(messages, tools)
    print(run1)

    # 延迟 7 秒
    time.sleep(7)

    # 将 user_query2 追加到消息历史记录
    messages.append(user_query2)

    # 运行 2：附加查询
    print("\n运行 2:") 
    run2 = completion_run(messages, tools)
    print(run2)


# 运行主函数
main(messages, tools, user_query2)

运行 1:
{
    "id": "50b96405-d8fb-4efb-8a2b-ffdb7923e5b4",
    "choices": [
        {
            "finish_reason": "tool_calls",
            "index": 0,
            "logprobs": null,
            "message": {
                "content": "",
                "role": "assistant",
                "tool_calls": [
                    {
                        "id": "call_0_aa4e917a-bf89-4ef0-a28a-5a66ddf6c229",
                        "function": {
                            "arguments": "{\"order_id\":\"9876543210\"}",
                            "name": "get_delivery_date"
                        },
                        "type": "function",
                        "index": 0
                    }
                ]
            }
        }
    ],
    "created": 1737820877,
    "model": "deepseek-chat",
    "object": "chat.completion",
    "system_fingerprint": "fp_3a5770e1b4",
    "usage": {
        "completion_tokens": 25,
        "prompt_tokens": 1037,
        "total_tokens": 1062,
       

对上述输出的解释：
运行1的解释：
id：

本次 API 调用的唯一标识符：50b96405-d8fb-4efb-8a2b-ffdb7923e5b4。

choices：

模型生成的响应结果。

finish_reason 为 tool_calls，表示模型决定调用外部工具（函数）来完成请求。

message 中的 tool_calls 是模型决定调用的函数列表。

tool_calls：

模型决定调用一个函数：get_delivery_date。

函数的参数为 {"order_id": "9876543210"}，表示查询订单 9876543210 的配送日期。

usage：

本次 API 调用的 token 使用情况：

prompt_tokens：1037，表示输入提示的 token 数量。

completion_tokens：25，表示模型生成的响应 token 数量。

total_tokens：1062，表示总 token 数量。

prompt_cache_hit_tokens 为 0，表示没有命中缓存。

运行2的解释：id：

本次 API 调用的唯一标识符：831e72ad-670d-4999-8a92-2eb96b20ea2b。

choices：

模型生成的响应结果。

finish_reason 为 tool_calls，表示模型决定调用外部工具（函数）来完成请求。

message 中的 tool_calls 是模型决定调用的函数列表。

tool_calls：

模型决定调用两个函数：

第一个函数：get_delivery_date，参数为 {"order_id": "9876543210"}，表示查询订单 9876543210 的配送日期。

第二个函数：cancel_order，参数为 {"order_id": "9876543210", "reason": "Decided to purchase locally for faster delivery"}，表示取消订单 9876543210，并提供了取消原因。

usage：

本次 API 调用的 token 使用情况：

prompt_tokens：1079，表示输入提示的 token 数量。

completion_tokens：57，表示模型生成的响应 token 数量。

total_tokens：1136，表示总 token 数量。

prompt_cache_hit_tokens 为 1024，表示有 1024 个 token 命中了缓存，减少了实际请求的 token 数量。

## 4.2 流式输出

In [312]:
import time
import json
from openai import OpenAI
from dotenv import load_dotenv
import os

# 加载 .env 文件
load_dotenv()

# 从环境变量中获取 API Key 和 Base URL
api_key = os.getenv("API_KEY")
base_url = os.getenv("BASE_URL")

# 初始化客户端
client = OpenAI(api_key=api_key, base_url=base_url)

In [314]:
stream = client.chat.completions.create(
    messages=[
        {
            "role": "user",
            "content": "有哪几种解决大模型幻觉的方法？",
        }
    ],
    model="deepseek-chat",
    max_tokens=300,
    stream=True
)
print("开始响应:")
for chunk in stream:
    print(chunk)

开始响应:
ChatCompletionChunk(id='49059b00-dacf-4f83-aeaf-b741fea9b610', choices=[Choice(delta=ChoiceDelta(content='', function_call=None, refusal=None, role='assistant', tool_calls=None), finish_reason=None, index=0, logprobs=None)], created=1737821722, model='deepseek-chat', object='chat.completion.chunk', service_tier=None, system_fingerprint='fp_3a5770e1b4', usage=None)
ChatCompletionChunk(id='49059b00-dacf-4f83-aeaf-b741fea9b610', choices=[Choice(delta=ChoiceDelta(content='大', function_call=None, refusal=None, role=None, tool_calls=None), finish_reason=None, index=0, logprobs=None)], created=1737821722, model='deepseek-chat', object='chat.completion.chunk', service_tier=None, system_fingerprint='fp_3a5770e1b4', usage=None)
ChatCompletionChunk(id='49059b00-dacf-4f83-aeaf-b741fea9b610', choices=[Choice(delta=ChoiceDelta(content='模型', function_call=None, refusal=None, role=None, tool_calls=None), finish_reason=None, index=0, logprobs=None)], created=1737821722, model='deepseek-chat', obj

In [315]:
stream = client.chat.completions.create(
    messages=[
        {
            "role": "user",
            "content": "有哪几种大模型微调的方法？",
        }
    ],
    model="deepseek-chat",
    max_tokens=600,
    stream=True
)
# 处理我们的流式输出
for chunk in stream:
    print(chunk)

ChatCompletionChunk(id='b637ad5a-c6cb-4dc3-9d5c-0eb02843e780', choices=[Choice(delta=ChoiceDelta(content='', function_call=None, refusal=None, role='assistant', tool_calls=None), finish_reason=None, index=0, logprobs=None)], created=1737821801, model='deepseek-chat', object='chat.completion.chunk', service_tier=None, system_fingerprint='fp_3a5770e1b4', usage=None)
ChatCompletionChunk(id='b637ad5a-c6cb-4dc3-9d5c-0eb02843e780', choices=[Choice(delta=ChoiceDelta(content='大', function_call=None, refusal=None, role=None, tool_calls=None), finish_reason=None, index=0, logprobs=None)], created=1737821801, model='deepseek-chat', object='chat.completion.chunk', service_tier=None, system_fingerprint='fp_3a5770e1b4', usage=None)
ChatCompletionChunk(id='b637ad5a-c6cb-4dc3-9d5c-0eb02843e780', choices=[Choice(delta=ChoiceDelta(content='模型', function_call=None, refusal=None, role=None, tool_calls=None), finish_reason=None, index=0, logprobs=None)], created=1737821801, model='deepseek-chat', object='c

In [316]:
import openai
import tiktoken


# 选择模型的编码方式，这里尽管基座是deepseek，但是编码用的gpt-4o-mini
encoding = tiktoken.encoding_for_model("gpt-4o-mini")

# 初始化令牌计数器
token_count = 0

# 创建流式聊天完成请求
response = client.chat.completions.create(
    model="deepseek-chat",
    messages=[{"role": "user", "content": "既然大模型已经足够强。为什么还需要微调和RAG？"}],
    stream=True
)

# 处理流式响应
for chunk in response:
    # 提取流式块中的内容部分
    if chunk.choices[0].delta.content:
        content_chunk =chunk.choices[0].delta.content
        # 计算当前内容块的令牌数量并累加
        token_count += len(encoding.encode(content_chunk))
        # 实时输出内容
        print(content_chunk, end="")

# 输出总的令牌使用数量
print(f"\n总使用Token数量（估）：{token_count}")

大模型（如GPT-3、GPT-4等）虽然在许多任务上表现出色，但它们仍然存在一些局限性，因此微调（Fine-tuning）和检索增强生成（Retrieval-Augmented Generation, RAG）等技术仍然非常重要。以下是几个主要原因：

### 1. **领域特定知识的不足**
   - **大模型的泛化能力**：大模型在训练时接触了大量的通用数据，因此在许多通用任务上表现良好。然而，对于某些特定领域（如医学、法律、金融等），大模型可能缺乏足够的专业知识。
   - **微调的作用**：通过在特定领域的数据上进行微调，模型可以更好地适应这些领域的任务，提升在该领域的表现。

### 2. **实时性和动态信息的缺失**
   - **大模型的静态知识**：大模型的知识截止于其训练数据的最后时间点，无法获取训练之后的最新信息。
   - **RAG的作用**：RAG通过结合外部知识库或实时检索系统，能够动态获取最新的信息，从而生成更准确和及时的响应。

### 3. **任务特定优化**
   - **通用模型的局限性**：虽然大模型在许多任务上表现良好，但对于某些特定任务（如情感分析、命名实体识别等），通用模型可能不如专门针对该任务进行微调的模型。
   - **微调的作用**：通过在特定任务的数据上进行微调，模型可以更好地适应这些任务的需求，提升任务性能。

### 4. **减少幻觉（Hallucination）**
   - **大模型的幻觉问题**：大模型有时会生成看似合理但实际上不准确或无依据的内容，这种现象被称为“幻觉”。
   - **RAG的作用**：RAG通过结合外部知识库，能够减少模型生成不准确信息的可能性，从而提高生成内容的准确性。

### 5. **提高效率和成本效益**
   - **大模型的资源消耗**：大模型通常需要大量的计算资源，运行成本较高。
   - **微调和RAG的作用**：通过微调和RAG，可以在保持较高性能的同时，减少对大规模模型的依赖，从而降低运行成本。

### 6. **个性化需求**
   - **用户或企业的个性化需求**：不同的用户或企业可能有不同的需求，通用模型可能无法完全满足这些个性化需求。
   - **微调的作用**：通过在特定用户或企业的数据上进行微调，模型可以更好地满足这些个性化需